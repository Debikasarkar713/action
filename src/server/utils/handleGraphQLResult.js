import jwtDecode from 'jwt-decode';

const setSocketAuth = (socket, authTokenStr) => {
  socket.setAuthToken({
    ...jwtDecode(authTokenStr),
    exp: undefined
  });
};

const handleGraphQLResult = (result, socket) => {
  const {data, errors} = result;
  if (errors) return false;
  if (data.notificationsAdded) {
    const {notifications} = data.notificationsAdded;
    const updateTokenNotification = notifications.find((notification) => notification.authToken);
    if (updateTokenNotification) {
      setSocketAuth(socket, updateTokenNotification.authToken);
      return true;
    }
  } else if (data.acceptTeamInviteNotification) {
    setSocketAuth(socket, data.acceptTeamInviteNotification.authToken);
    return true;
  } else if (data.newAuthToken) {
    // TODO refactor the above 2 clauses so this is the only one
    setSocketAuth(socket, data.newAuthToken);
    return true;
  }
  return false;
};

export default handleGraphQLResult;
