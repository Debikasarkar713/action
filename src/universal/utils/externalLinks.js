// eslint-disable-next-line import/prefer-default-export
export const PRICING_LINK = 'https://www.parabol.co/pricing/';
