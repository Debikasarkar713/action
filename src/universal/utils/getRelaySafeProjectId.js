import fromGlobalId from 'universal/utils/relay/fromGlobalId';

const getRelaySafeProjectId = (projectId) => {
  try {
    const {id: globalId} = fromGlobalId(projectId);
    return globalId;
  } catch (e) {
    return projectId;
  }
};
export default getRelaySafeProjectId;
