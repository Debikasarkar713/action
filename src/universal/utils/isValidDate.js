const isValidDate = (maybeDate) => {
  return !isNaN(maybeDate.getTime());
};

export default isValidDate;
