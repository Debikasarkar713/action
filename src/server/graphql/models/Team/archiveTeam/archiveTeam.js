import {GraphQLBoolean, GraphQLNonNull, GraphQLString} from 'graphql';
import getRethink from 'server/database/rethinkDriver';
import {auth0ManagementClient} from 'server/utils/auth0Helpers';
import {getUserId, requireSUOrLead, requireWebsocket} from 'server/utils/authorization';
import getPubSub from 'server/utils/getPubSub';
import sendSegmentEvent from 'server/utils/sendSegmentEvent';
import tmsSignToken from 'server/utils/tmsSignToken';
import shortid from 'shortid';
import {NEW_AUTH_TOKEN, NOTIFICATIONS_ADDED, TEAM_ARCHIVED} from 'universal/utils/constants';

export default {
  type: GraphQLBoolean,
  args: {
    teamId: {
      type: new GraphQLNonNull(GraphQLString),
      description: 'The teamId to archive (or delete, if team is unused)'
    }
  },
  async resolve(source, {teamId}, {authToken, socket}) {
    const r = getRethink();
    const now = new Date();

    // AUTH
    const userId = getUserId(authToken);
    const teamMemberId = `${userId}::${teamId}`;
    requireSUOrLead(authToken, teamMemberId);
    requireWebsocket(socket);

    // RESOLUTION
    sendSegmentEvent('Archive Team', userId, {teamId});
    const {teamResults: {notificationData}, userDocs} = await r.table('Team')
      .get(teamId)
      .pluck('name', 'orgId')
      .do((team) => ({
        projectCount: r.table('Project').getAll(teamId, {index: 'teamId'}).count(),
        team,
        userIds: r.table('TeamMember')
          .getAll(teamId, {index: 'teamId'})
          .filter({isNotRemoved: true})('userId')
          .coerceTo('array')
      }))
      .do((doc) => ({
        userDocs: r.table('User')
          .getAll(r.args(doc('userIds')), {index: 'id'})
          .update((user) => ({tms: user('tms').difference([teamId])}), {returnChanges: true})('changes')('new_val')
          .pluck('id', 'tms')
          .default([]),
        teamName: doc('team')('name'),
        teamResults: r.branch(
          r.and(doc('projectCount').eq(0), doc('userIds').count().eq(1)),
          {
            // Team has no projects nor addn'l TeamMembers, hard delete it:
            teamResult: r.table('Team').get(teamId).delete(),
            teamMemberResult: r.table('TeamMember').getAll(teamId, {index: 'teamId'}).delete()
          },
          {
            notificationData: doc,
            teamResult: r.table('Team').get(teamId).update({isArchived: true})
          }
        )
      }));

    if (notificationData) {
      const {team: {name: teamName, orgId}, userIds} = notificationData;
      const notifications = userIds.map((notifiedUserId) => ({
        id: shortid.generate(),
        orgId,
        startAt: now,
        type: TEAM_ARCHIVED,
        userIds: [notifiedUserId],
        teamName
      }));
      await r.table('Notification').insert(notifications);

      notifications.forEach((notification) => {
        const notificationsAdded = {notifications: [notification]};
        getPubSub().publish(`${NOTIFICATIONS_ADDED}.${notification.userIds[0]}`, {notificationsAdded});
      });
    }
    userDocs.forEach((user) => {
      const {id, tms} = user;
      // update the tms on auth0 in async
      auth0ManagementClient.users.updateAppMetadata({id}, {tms});
      // update the server socket, if they're logged in
      getPubSub().publish(`${NEW_AUTH_TOKEN}.${id}`, {newAuthToken: tmsSignToken({sub: userId}, tms)});
    });

    return true;
  }
};
