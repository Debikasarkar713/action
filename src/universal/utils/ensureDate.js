import isValidDate from 'universal/utils/isValidDate';

const ensureDate = (maybeDate) => {
  const date = new Date(maybeDate);
  return isValidDate(date) ? date : new Date();
};

export default ensureDate;
