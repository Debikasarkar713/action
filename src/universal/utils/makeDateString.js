import ensureDate from 'universal/utils/ensureDate';

export const days = [
  'Sunday',
  'Monday',
  'Tuesday',
  'Wednesday',
  'Thursday',
  'Friday',
  'Saturday'
];

export const months = [
  'January',
  'February',
  'March',
  'April',
  'May',
  'June',
  'July',
  'August',
  'September',
  'October',
  'November',
  'December'
];

export default function makeDateString(datetime, options = {}) {
  const timestamp = ensureDate(datetime);
  const {showDay} = options;
  const day = timestamp.getDay();
  const month = timestamp.getMonth();
  const date = timestamp.getDate();
  const year = timestamp.getFullYear();
  const dayPart = showDay ? `${days[day]}, ` : '';
  return `${dayPart}${months[month]} ${date}, ${year}`;
}
