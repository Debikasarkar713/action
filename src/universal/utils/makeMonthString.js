import {months} from 'universal/utils/makeDateString';
import ensureDate from 'universal/utils/ensureDate';

export default function makeMonthString(datetime) {
  const timestamp = ensureDate(datetime);
  const month = timestamp.getMonth();
  const year = timestamp.getFullYear();
  return `${months[month]} ${year}`;
}
