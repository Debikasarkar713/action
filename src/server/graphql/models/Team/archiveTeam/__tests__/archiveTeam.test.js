import DynamicSerializer from 'dynamic-serializer';
import MockDate from 'mockdate';
import getRethink from 'server/database/rethinkDriver';
import {__now} from 'server/__tests__/setup/mockTimes';
import fetchAndSerialize from 'server/__tests__/utils/fetchAndSerialize';
import MockDB from 'server/__tests__/setup/MockDB';
import archiveTeam from 'server/graphql/models/Team/archiveTeam/archiveTeam';
import mockAuthToken from 'server/__tests__/setup/mockAuthToken';
import exchange from 'server/__mocks__/exchange';
import socket from 'server/__mocks__/socket';
import {auth0ManagementClient} from 'server/utils/auth0Helpers';
import MockPubSub from 'server/__mocks__/MockPubSub';
import * as getPubSub from 'server/utils/getPubSub';
import * as tmsSignToken from 'server/utils/tmsSignToken';

MockDate.set(__now);
console.error = jest.fn();

describe('ArchiveTeam', () => {
  test('archives a team', async () => {
    // SETUP
    const r = getRethink();
    const dynamicSerializer = new DynamicSerializer();
    const mockPubSub = new MockPubSub();
    const mockDB = new MockDB();
    const {user, team: [updatedTeam], teamMember} = await mockDB.init();
    updatedTeam.isArchived = true;
    const teamLeadId = teamMember.find((tm) => tm.teamId === updatedTeam.id && tm.isLead).userId;
    const teamLead = user.find((usr) => usr.id === teamLeadId);
    const authToken = mockAuthToken(teamLead);
    auth0ManagementClient.__initMock(mockDB.db);
    auth0ManagementClient.users.updateAppMetadata.mockReset();
    getPubSub.default = () => mockPubSub;
    tmsSignToken.default = jest.fn(() => 'FAKEENCODEDJWT');
    // TEST
    await archiveTeam.resolve(undefined, {teamId: updatedTeam.id}, {authToken, exchange, socket});

    // VERIFY
    const db = await fetchAndSerialize({
      notification: r.table('Notification').getAll(updatedTeam.orgId, {index: 'orgId'}).orderBy('startAt'),
      team: r.table('Team').get(updatedTeam.id)
    }, dynamicSerializer);
    expect(db).toMatchSnapshot();
    expect(auth0ManagementClient.users.updateAppMetadata.mock.calls.length)
      .toBe(teamMember.length);
    expect(mockPubSub.__serialize(dynamicSerializer)).toMatchSnapshot();
  });

  test('deletes a team when it has no projects or other team members', async () => {
    // SETUP
    const r = getRethink();
    const dynamicSerializer = new DynamicSerializer();
    const mockPubSub = new MockPubSub();
    const mockDB = new MockDB();
    const {user: [user], team: [updatedTeam]} = await mockDB
      .newOrg({name: 'Sad Sacks, Inc.'})
      .newTeam({name: 'The Lonely Ones'})
      .newUser({name: 'Leader of One'})
      .newTeamMember({isLead: true});
    updatedTeam.isArchived = true;
    const authToken = mockAuthToken(user);
    auth0ManagementClient.__initMock(mockDB.db);
    auth0ManagementClient.users.updateAppMetadata.mockReset();
    exchange.publish.mockReset();
    getPubSub.default = () => mockPubSub;
    tmsSignToken.default = jest.fn(() => 'FAKEENCODEDJWT');
    // TEST
    await archiveTeam.resolve(undefined, {teamId: updatedTeam.id}, {authToken, exchange, socket});

    // VERIFY
    const db = await fetchAndSerialize({
      notification: r.table('Notification').getAll(updatedTeam.orgId, {index: 'orgId'}).orderBy('startAt'),
      team: r.table('Team').get(updatedTeam.id).default({}),
      teamMember: r.table('TeamMember').getAll(updatedTeam.id, {index: 'teamId'}).default([]),
      user: r.table('User').get(user.id)
    }, dynamicSerializer);
    expect(db).toMatchSnapshot();
    expect(auth0ManagementClient.users.updateAppMetadata.mock.calls.length)
      .toBe(1);
    expect(mockPubSub.__serialize(dynamicSerializer)).toMatchSnapshot();
  });
});
