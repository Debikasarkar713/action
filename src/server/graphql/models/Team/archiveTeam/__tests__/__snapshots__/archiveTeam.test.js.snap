// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ArchiveTeam archives a team 1`] = `
Object {
  "notification": Array [
    Object {
      "id": 0,
      "orgId": 1,
      "startAt": 2017-01-01T08:00:00.000Z,
      "teamName": "Team America",
      "type": "TEAM_ARCHIVED",
      "userIds": Array [
        2,
        3,
        4,
        5,
        6,
        7,
        8,
        9,
      ],
    },
  ],
  "team": Object {
    "activeFacilitator": null,
    "facilitatorPhase": "lobby",
    "facilitatorPhaseItem": null,
    "id": 10,
    "isArchived": true,
    "isPaid": true,
    "meetingId": null,
    "meetingPhase": "lobby",
    "meetingPhaseItem": null,
    "name": "Team America",
    "orgId": 1,
    "tier": "personal",
  },
}
`;

exports[`ArchiveTeam archives a team 2`] = `
Object {
  "notificationsAdded": Array [
    Object {
      "channelId": 10,
      "message": Object {
        "notificationsAdded": Object {
          "notifications": Array [
            Object {
              "id": 0,
              "orgId": 1,
              "startAt": 2017-01-01T08:00:00.000Z,
              "teamName": "Team America",
              "type": "TEAM_ARCHIVED",
              "userIds": Array [
                2,
                3,
                4,
                5,
                6,
                7,
                8,
                9,
              ],
            },
          ],
        },
      },
    },
    Object {
      "channelId": 9,
      "message": Object {
        "notificationsAdded": Object {
          "notifications": Array [
            Object {
              "authToken": 11,
              "type": "KICKED_OUT",
            },
          ],
        },
      },
    },
    Object {
      "channelId": 8,
      "message": Object {
        "notificationsAdded": Object {
          "notifications": Array [
            Object {
              "authToken": 11,
              "type": "KICKED_OUT",
            },
          ],
        },
      },
    },
    Object {
      "channelId": 7,
      "message": Object {
        "notificationsAdded": Object {
          "notifications": Array [
            Object {
              "authToken": 11,
              "type": "KICKED_OUT",
            },
          ],
        },
      },
    },
    Object {
      "channelId": 6,
      "message": Object {
        "notificationsAdded": Object {
          "notifications": Array [
            Object {
              "authToken": 11,
              "type": "KICKED_OUT",
            },
          ],
        },
      },
    },
    Object {
      "channelId": 5,
      "message": Object {
        "notificationsAdded": Object {
          "notifications": Array [
            Object {
              "authToken": 11,
              "type": "KICKED_OUT",
            },
          ],
        },
      },
    },
    Object {
      "channelId": 4,
      "message": Object {
        "notificationsAdded": Object {
          "notifications": Array [
            Object {
              "authToken": 11,
              "type": "KICKED_OUT",
            },
          ],
        },
      },
    },
    Object {
      "channelId": 3,
      "message": Object {
        "notificationsAdded": Object {
          "notifications": Array [
            Object {
              "authToken": 11,
              "type": "KICKED_OUT",
            },
          ],
        },
      },
    },
    Object {
      "channelId": 2,
      "message": Object {
        "notificationsAdded": Object {
          "notifications": Array [
            Object {
              "authToken": 11,
              "type": "KICKED_OUT",
            },
          ],
        },
      },
    },
  ],
}
`;

exports[`ArchiveTeam deletes a team when it has no projects or other team members 1`] = `
Object {
  "notification": Array [],
  "team": Object {},
  "teamMember": Array [],
  "user": Object {
    "cachedAt": 2017-01-01T07:00:00.000Z,
    "createdAt": 2017-01-01T07:00:00.000Z,
    "emailVerified": false,
    "id": 0,
    "identities": Array [],
    "inactive": false,
    "lastLogin": 2017-01-01T07:00:00.000Z,
    "lastSeenAt": 2017-01-01T07:00:00.000Z,
    "name": "Leader of One",
    "picture": "https://placeimg.com/100/100/animals",
    "preferredName": "Leader of One",
    "tms": Array [],
    "updatedAt": 2017-01-01T07:00:00.000Z,
    "userOrgs": Array [
      Object {
        "id": 1,
        "role": null,
      },
    ],
    "welcomeSentAt": 2017-01-01T07:00:00.000Z,
  },
}
`;

exports[`ArchiveTeam deletes a team when it has no projects or other team members 2`] = `
Object {
  "notificationsAdded": Array [
    Object {
      "channelId": 0,
      "message": Object {
        "notificationsAdded": Object {
          "notifications": Array [
            Object {
              "authToken": 2,
              "type": "KICKED_OUT",
            },
          ],
        },
      },
    },
  ],
}
`;
